server/src/main/java/org/opensearch/repositories/RepositoriesService.java

```diff
@@ -763,6 +777,86 @@ public void updateRepositoriesMap(Map<String, Repository> repos) {
+
+                if (segmentRepositoryInIndex.equals(repository) || translogRepositoryInIndex.equals(repository)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    private static boolean isSystemRepositorySettingPresent(Settings repositoryMetadataSettings) {
+        return SYSTEM_REPOSITORY_SETTING.get(repositoryMetadataSettings);
+    }
+
+    private static boolean isValueEqual(String newValue, String currentValue) {
+        if (newValue == null) {
+            throw new IllegalArgumentException("new value cannot be empty, " + "current value [" + currentValue + "]");
+        }
+        if (newValue.equals(currentValue) == false) {
+            throw new IllegalArgumentException(
+                "trying to modify an unmodifiable attribute of system repository from "
+                    + "current value ["
+                    + currentValue
+                    + "] to new value ["
+                    + newValue
+                    + "]"
+            );
+        }
+        return true;
+    }
+
+    public RepositoryMetadata ensureValidSystemRepositoryUpdate(
+        RepositoryMetadata newRepositoryMetadata,
+        RepositoryMetadata currentRepositoryMetadata
+    ) {
+        if (isSystemRepositorySettingPresent(currentRepositoryMetadata.settings())) {
+            Settings.Builder updatedSettings = Settings.builder().put(newRepositoryMetadata.settings());
+            try {
+                isValueEqual(newRepositoryMetadata.type(), currentRepositoryMetadata.type());
+
+                Repository repository = repositories.get(currentRepositoryMetadata.name());
+                Settings newRepositoryMetadataSettings = newRepositoryMetadata.settings();
+                Settings currentRepositoryMetadataSettings = currentRepositoryMetadata.settings();
+
+                List<String> restrictedSettings = repository.getRestrictedSystemRepositorySettings()
+                    .stream()
+                    .map(setting -> setting.getKey())
+                    .collect(Collectors.toList());
+
+                for (String restrictedSettingKey : restrictedSettings) {
+                    String newSettingValue = newRepositoryMetadataSettings.get(restrictedSettingKey);
+                    String currentSettingValue = currentRepositoryMetadataSettings.get(restrictedSettingKey);
+
+                    isValueEqual(newSettingValue, currentSettingValue);
+
+                    if (currentSettingValue != null) {
+                        updatedSettings.put(restrictedSettingKey, currentSettingValue);
+                    }
+                }
+            } catch (IllegalArgumentException e) {
+                throw new RepositoryException(currentRepositoryMetadata.name(), e.getMessage());
+            }
+            return new RepositoryMetadata(currentRepositoryMetadata.name(), currentRepositoryMetadata.type(), updatedSettings.build());
```

Comment:
include `cryptoMetadata` as well .

Reply:
this should not return anything and just validate  . Rename to `validateSystemRepositoryUpdate`